assets/src/edit-story/app/story/actions/useAutoSave.js

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+import { useCallback, useState } from 'react';
+import { renderToStaticMarkup } from 'react-dom/server';
+
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import objectPick from '../../../utils/objectPick';
+import { useAPI } from '../../api';
+import { useConfig } from '../../config';
+import OutputStory from '../../../output/story';
+import { useSnackbar } from '../../snackbar';
+
+/**
+ * Creates AMP HTML markup for saving to DB for rendering in the FE.
+ *
+ * @param {import('../../../types').Story} story Story object.
+ * @param {Array<Object>} pages List of pages.
+ * @param {Object} metadata Metadata.
+ * @return {Element} Story markup.
+ */
+const getStoryMarkup = (story, pages, metadata) => {
+  return renderToStaticMarkup(
+    <OutputStory story={story} pages={pages} metadata={metadata} />
+  );
+};
+
+/**
+ * Custom hook to auto-save a story.
+ *
+ * @param {Object}    properties Properties to update.
+ * @param {number}    properties.storyId Story post id.
+ * @param {Array}     properties.pages Array of all pages.
+ * @param {Object}    properties.story Story-global properties
+ * @return {Function} Function that can be called to save a story.
+ */
+function useAutoSave({ storyId, pages, story }) {
+  const {
+    actions: { autoSaveById },
+  } = useAPI();
+  const { metadata } = useConfig();
+  const { showSnackbar } = useSnackbar();
+  const [isAutoSaving, setIsAutoSaving] = useState(false);
+
+  const autoSave = useCallback(
+    (props) => {
+      setIsAutoSaving(true);
+      const propsToSave = objectPick(story, [
+        'title',
+        'status',
+        'author',
+        'date',
+        'modified',
+        'slug',
+        'excerpt',
+        'featuredMedia',
+        'password',
+        'publisherLogo',
+        'stylePresets',
+        'autoAdvance',
+        'defaultPageDuration',
+      ]);
+      const content = getStoryMarkup(story, pages, metadata);
+      return autoSaveById({
+        storyId,
+        content,
+        pages,
+        ...propsToSave,
+        ...props,
+      })
+        .catch(() => {
+          showSnackbar({
+            message: __(
+              'Failed to generate preview for the story',
```

Comment:
I suppose we don't actually need this since we already have the dialog opening with the option to try again if opening the preview fails. ?

Reply:
Seems like it. Once we implement periodic autosaving, it would be odd to get such a message every 60 seconds.

Since we already know at this point that autosaving failed, perhaps this could be an opportunity to show the dialog sooner rather than first waiting for 5 seconds.